### PR TITLE
Feature/dqn

### PR DESCRIPTION
**Pull request by akshatbjain**
Merging the DQN branch to develop!

### CODE REVIEW

File: drl/nn/nn.py [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+		if self.network_type == 'conv':
+			if options_in is not None:
+				convnet_options.update(options_in)
+			self.options = convnet_options
+		else:
+			if options_in is not None:
+				nn_options.update(options_in)
+			self.options = nn_options
+
+		if self.mode == 'test':
```
Comment (BartKeulen) [inline drl/nn/nn.py]: How does this work in practice? The reinforcement learning cycle consists out of taking a step (inference/test) followed by updating parameters (train). I think it should have full access during a step and be in test mode during updating. 

I had a similar problem when implementing batch normalization, when you take a step you only have one data point, if you are still in train mode your single observation will become zero all the time. See in utilities/tfnetwork the bn_layer function for how i implemented this using a tf.placeholder.

**Review by BartKeulen — COMMENTED** [review]
Looks good! I have added one comment. Please come and see me so i can give a little bit better explanation.

File: drl/nn/nn.py [hunk 2]
```diff
@@ -0,0 +1,453 @@
...
+			:param epsilon: Small float added to variance to avoid dividing by zero.
+			:param name: Optional Parameter. Allows you to name the layer.
+			:return: Batch normalized layer output.
+		"""
+		scale = tf.Variable(tf.ones([input_layer.get_shape()[-1]]))
+		beta = tf.Variable(tf.zeros([input_layer.get_shape()[-1]]))
+		pop_mean = tf.Variable(tf.zeros([input_layer.get_shape()[-1]]), trainable=False)
+		pop_var = tf.Variable(tf.ones([input_layer.get_shape()[-1]]), trainable=False)
+
+		if self.is_training:
```
Comment (BartKeulen) [inline drl/nn/nn.py]: You cannot use a tf.placeholder(bool) in an if statement. You have to implement like [this](https://github.com/BartKeulen/drl/blob/feature/ddpgSmartStart/drl/utilities/tfnetwork.py#L117)

File: drl/nn/nn.py [hunk 3]
```diff
@@ -0,0 +1,453 @@
...
+
+	def dropout(self, input_layer, name=None):
+		"""
+		Returns dropout with user-defined keep_prob if training. Uses keep_prob of 1.0 if testing.
+
+			:param input_layer Input layer to which we will apply dropout
+			:param name Optional Parameter. Allows you to name the layer.
+			:return: dropout layer
+		"""
+		if self.is_training:
```
Comment (BartKeulen) [inline drl/nn/nn.py]: Same as previous comment

File: drl/nn/nn.py (hunk 2 above)
Comment (akshatbjain) [reply drl/nn/nn.py]: Yes saw that. Working on it :)

File: drl/nn/nn.py [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+			                      pop_var * decay + batch_var * (1 - decay))
+			with tf.control_dependencies([train_mean, train_var]):
+				return tf.nn.batch_normalization(input_layer,
+				                                 batch_mean, batch_var, beta, scale, epsilon, name=name)
+
+		def test():
+			return tf.nn.batch_normalization(input_layer,
+			                                 pop_mean, pop_var, beta, scale, epsilon, name=name)
+
+		return tf.cond(self.is_training, lambda: train(), lambda: test())
```
Comment (BartKeulen) [inline drl/nn/nn.py]: Probably the reason why it is not working without lambda is because you pass the called function, e.g. look at this example:

`tf.cond(self.is_training, train(), test())`
What will happen is, the `train` and `test` function are called and the results (which is `tf.nn.batch_normalization(...)`) are passed into the `tf.cond` function, and the results cannot be called.

Now take the this example:
`tf.cond(self.is_training, train, test)`
In this example the functions `train` and `test` it self is passed into the `tf.cond` function and within the `tf.cond` function some code will execute them by calling `train()` or `test()` at some point.

The reason why `lamda: train()` works is because you construct a new function definition which is passed into `tf.cond`. You can see it as this `new_func = lambda: train()` and `new_func` is passed into `tf_cond`. If you would pass `new_func()` you would get the same problem as the before.

I hope this is clear, i have noticed that passing functions into other functions is a really powerful tool in Python, so it is good to know how it works.

Comment (BartKeulen) [inline drl/nn/nn.py]: Probably the reason why it is not working without lambda is because you pass the called function, e.g. look at this example:

`tf.cond(self.is_training, train(), test())`
What will happen is, the `train` and `test` function are called and the results (which is `tf.nn.batch_normalization(...)`) are passed into the `tf.cond` function, E.g. lets say `result = train()` and now `result` is passed into `tf.cond`, but you cannot perform `result()`. 

Now take the this example:
`tf.cond(self.is_training, train, test)`
In this example the functions `train` and `test` it self is passed into the `tf.cond` function and within the `tf.cond` function some code will execute them by calling `train()` or `test()` at some point. E.g. `train` is passed into `tf.cond` and when needed the function is executed to return `result = train()`.

The reason why `lamda: train()` works is because you construct a new function definition which is passed into `tf.cond`. You can see it as this `new_func = lambda: train()` and `new_func` is passed into `tf_cond`. If you would pass `new_func()` you would get the same problem as the before.

I hope this is clear, i have noticed that passing functions into other functions is a really powerful tool in Python, so it is good to know how it works.
